File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/patient/assessments/SEUrinalysis.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
package mil.tatrc.physiology.datamodel.patient.assessments;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.EnumClarityIndicator;
import mil.tatrc.physiology.datamodel.bind.EnumPresenceIndicator;
import mil.tatrc.physiology.datamodel.bind.EnumUrineColor;
import mil.tatrc.physiology.datamodel.bind.UrinalysisData;
import mil.tatrc.physiology.datamodel.properties.*;

public class SEUrinalysis extends SEPatientAssessment
{
  protected EnumUrineColor          color;
  protected EnumClarityIndicator    apperance;
  protected EnumPresenceIndicator   glucose;
  protected EnumPresenceIndicator   ketone;
  protected SEScalar                bilirubin;
  protected SEScalar                specificGravity;
  protected EnumPresenceIndicator   blood;
  protected SEScalar                pH;  
  protected EnumPresenceIndicator   protein;
  protected SEScalarMassPerVolume   urobilinogen;
  protected EnumPresenceIndicator   nitrite;
  protected EnumPresenceIndicator   leukocyteEsterase;
  
  public void clear()
  {
    super.clear();
    this.color = null;
    this.apperance = null;
    this.urobilinogen = null;
    this.ketone = null;
    this.bilirubin = null;
    this.specificGravity = null;
    this.blood = null;
    this.pH = null;
    this.protein = null;
    this.urobilinogen = null;
    this.nitrite = null;
    this.leukocyteEsterase = null;
  }
  
  public void reset()
  {
    super.reset();
    this.color = null;
    this.apperance = null;
    this.urobilinogen = null;
    this.ketone = null;
    if(this.bilirubin!=null)
      this.bilirubin.invalidate();
    if(this.specificGravity!=null)
      this.specificGravity.invalidate();    
    this.blood = null;
    if(this.pH!=null)
      this.pH.invalidate();    
    this.protein = null;
    if(this.urobilinogen!=null)
      this.urobilinogen.invalidate();    
    this.nitrite = null;
    this.leukocyteEsterase = null;   
  }
  
  public boolean load(UrinalysisData in)
  {
    super.load(in);
    if(in.getColor()!=null)
      this.setColor(in.getColor());
    if(in.getAppearance()!=null)
      this.setAppearance(in.getAppearance());
    if(in.getGlucose()!=null)
      this.setGlucose(in.getGlucose());
    if(in.getKetone()!=null)
      this.setKetone(in.getKetone());
    if(in.getBilirubin()!=null)
      this.getBilirubin().load(in.getBilirubin());
    if(in.getSpecificGravity()!=null)
      this.getSpecificGravity().load(in.getSpecificGravity());
    if(in.getBlood()!=null)
      this.setBlood(in.getBlood());
    if(in.getPH()!=null)
      this.getPH().load(in.getPH());
    if(in.getProtein()!=null)
      this.setProtein(in.getProtein());
    if(in.getUrobilinogen()!=null)
      this.getUrobilinogen().load(in.getUrobilinogen());
    if(in.getNitrite()!=null)
      this.setNitrite(in.getNitrite());
    if(in.getLeukocyteEsterase()!=null)
      this.setLeukocyteEsterase(in.getLeukocyteEsterase());
    return true;
  }
  
  public UrinalysisData unload()
  {
    UrinalysisData data = CDMSerializer.objFactory.createUrinalysisData();
    unload(data);
    return data;
  }
  
  protected void unload(UrinalysisData data)
  {
    super.unload(data);
    if (color != null)
      data.setColor(color);
    if (apperance != null)
      data.setAppearance(apperance);
    if (glucose != null)
      data.setGlucose(glucose);
    if (ketone != null)
      data.setKetone(ketone);
    if (bilirubin != null)
      data.setBilirubin(bilirubin.unload());
    if (specificGravity != null)
      data.setSpecificGravity(specificGravity.unload());
    if (blood != null)
      data.setBlood(blood);
    if (pH != null)
      data.setPH(pH.unload());
    if (protein != null)
      data.setProtein(protein);
    if (urobilinogen != null)
      data.setUrobilinogen(urobilinogen.unload());
    if (nitrite != null)
      data.setNitrite(nitrite);
    if (leukocyteEsterase != null)
      data.setLeukocyteEsterase(leukocyteEsterase);
  }
  
  public EnumUrineColor  getColor() { return this.color;}
  public void    setColor(EnumUrineColor color){this.color=color;}
  public boolean hasColor(){return this.color==null?false:true;}
  
  public EnumClarityIndicator  getAppearance() { return this.apperance;}
  public void    setAppearance(EnumClarityIndicator a){this.apperance=a;}
  public boolean hasAppearance(){return this.apperance==null?false:true;}
  
  public EnumPresenceIndicator  getGlucose() { return this.glucose;}
  public void    setGlucose(EnumPresenceIndicator g){this.glucose=g;}
  public boolean hasGlucose(){return this.glucose==null?false:true;}
  
  public EnumPresenceIndicator  getKetone() { return this.ketone;}
  public void    setKetone(EnumPresenceIndicator k){this.ketone=k;}
  public boolean hasKetone(){return this.ketone==null?false:true;}
  
  public boolean hasBilirubin()
  {
    return bilirubin == null ? false : bilirubin.isValid();
  }
  public SEScalar getBilirubin()
  {
    if (bilirubin == null)
      bilirubin = new SEScalar();
    return bilirubin;
  }
  
  public boolean hasSpecificGravity()
  {
    return specificGravity == null ? false : specificGravity.isValid();
  }
  public SEScalar getSpecificGravity()
  {
    if (specificGravity == null)
      specificGravity = new SEScalar();
    return specificGravity;
  }
  
  public EnumPresenceIndicator  getBlood() { return this.blood;}
  public void    setBlood(EnumPresenceIndicator b){this.blood=b;}
  public boolean hasBlood(){return this.blood==null?false:true;}
  
  public boolean hasPH()
  {
    return pH == null ? false : pH.isValid();
  }
  public SEScalar getPH()
  {
    if (pH == null)
      pH = new SEScalar();
    return pH;
  }
  
  public EnumPresenceIndicator  getProtein() { return this.protein;}
  public void    setProtein(EnumPresenceIndicator p){this.protein=p;}
  public boolean hasProtein(){return this.protein==null?false:true;}
 
  
  public boolean hasUrobilinogen()
  {
    return urobilinogen == null ? false : urobilinogen.isValid();
  }
  public SEScalarMassPerVolume getUrobilinogen()
  {
    if (urobilinogen == null)
      urobilinogen = new SEScalarMassPerVolume();
    return urobilinogen;
  }
  
  public EnumPresenceIndicator  getNitrite() { return this.nitrite;}
  public void    setNitrite(EnumPresenceIndicator n){this.nitrite=n;}
  public boolean hasNitrite(){return this.nitrite==null?false:true;}
  
  public EnumPresenceIndicator  getLeukocyteEsterase() { return this.leukocyteEsterase;}
  public void    setLeukocyteEsterase(EnumPresenceIndicator l){this.leukocyteEsterase=l;}
  public boolean hasLeukocyteEsterase(){return this.leukocyteEsterase==null?false:true;}
}

File: projects/biogears/libBiogears/src/cdm/patient/assessments/SEUrinalysis.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/patient/assessments/SEUrinalysis.h>

#include <biogears/cdm/patient/SEPatient.h>
#include <biogears/cdm/properties/SEScalarMassPerTime.h>
#include <biogears/cdm/properties/SEScalarMassPerVolume.h>
#include <biogears/cdm/substance/SESubstanceManager.h>
#include <biogears/cdm/utils/GeneralMath.h>
#include "io/cdm/PatientAssessments.h"

namespace biogears {
SEUrinalysis::SEUrinalysis()
{
  m_Color = SEUrineColor::Invalid;
  m_Appearance = SEClarityIndicator::Invalid;
  m_Glucose = SEPresenceIndicator::Invalid;
  m_Ketone = SEPresenceIndicator::Invalid;
  m_Bilirubin = nullptr;
  m_SpecificGravity = nullptr;
  m_Blood = SEPresenceIndicator::Invalid;
  m_pH = nullptr;
  m_Protein = SEPresenceIndicator::Invalid;
  m_Urobilinogen = nullptr;
  m_Nitrite = SEPresenceIndicator::Invalid;
  m_LeukocyteEsterase = SEPresenceIndicator::Invalid;

  m_Microscopic = nullptr;
}
//-------------------------------------------------------------------------------
SEUrinalysis::~SEUrinalysis()
{
  Clear();
}
//-------------------------------------------------------------------------------
void SEUrinalysis::Clear()
{
  SEPatientAssessment::Clear();
  m_Color = SEUrineColor::Invalid;
  m_Appearance = SEClarityIndicator::Invalid;
  m_Glucose = SEPresenceIndicator::Invalid;
  m_Ketone = SEPresenceIndicator::Invalid;
  SAFE_DELETE(m_Bilirubin);
  SAFE_DELETE(m_SpecificGravity);
  m_Blood = SEPresenceIndicator::Invalid;
  SAFE_DELETE(m_pH);
  m_Protein = SEPresenceIndicator::Invalid;
  SAFE_DELETE(m_Urobilinogen);
  m_Nitrite = SEPresenceIndicator::Invalid;
  m_LeukocyteEsterase = SEPresenceIndicator::Invalid;

  SAFE_DELETE(m_Microscopic);
}
//-------------------------------------------------------------------------------
void SEUrinalysis::Reset()
{
  SEPatientAssessment::Reset();
  m_Color = SEUrineColor::Invalid;
  m_Appearance = SEClarityIndicator::Invalid;
  m_Glucose = SEPresenceIndicator::Invalid;
  m_Ketone = SEPresenceIndicator::Invalid;
  INVALIDATE_PROPERTY(m_Bilirubin);
  INVALIDATE_PROPERTY(m_SpecificGravity);
  m_Blood = SEPresenceIndicator::Invalid;
  INVALIDATE_PROPERTY(m_pH);
  m_Protein = SEPresenceIndicator::Invalid;
  INVALIDATE_PROPERTY(m_Urobilinogen);
  m_Nitrite = SEPresenceIndicator::Invalid;
  m_LeukocyteEsterase = SEPresenceIndicator::Invalid;

  SAFE_DELETE(m_Microscopic);
}

//-------------------------------------------------------------------------------
bool SEUrinalysis::HasColorResult() const
{
  return m_Color != SEUrineColor::Invalid;
}
SEUrineColor SEUrinalysis::GetColorResult() const
{
  return m_Color;
}
void SEUrinalysis::SetColorResult(SEUrineColor color)
{
  m_Color = color;
}
void SEUrinalysis::InvalidateColorResult()
{
  m_Color = SEUrineColor::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasAppearanceResult() const
{
  return m_Appearance != SEClarityIndicator::Invalid;
}
SEClarityIndicator SEUrinalysis::GetAppearanceResult() const
{
  return m_Appearance;
}
void SEUrinalysis::SetAppearanceResult(SEClarityIndicator c)
{
  m_Appearance = c;
}
void SEUrinalysis::InvalidateAppearanceResult()
{
  m_Appearance = SEClarityIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasGlucoseResult() const
{
  return m_Glucose != SEPresenceIndicator::Invalid;
}
SEPresenceIndicator SEUrinalysis::GetGlucoseResult() const
{
  return m_Glucose;
}
void SEUrinalysis::SetGlucoseResult(SEPresenceIndicator p)
{
  m_Glucose = p;
}
void SEUrinalysis::InvalidateGlucoseResult()
{
  m_Glucose = SEPresenceIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasKetoneResult() const
{
  return m_Ketone != SEPresenceIndicator::Invalid;
}
SEPresenceIndicator SEUrinalysis::GetKetoneResult() const
{
  return m_Ketone;
}
void SEUrinalysis::SetKetoneResult(SEPresenceIndicator p)
{
  m_Ketone = p;
}
void SEUrinalysis::InvalidateKetoneResult()
{
  m_Ketone = SEPresenceIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasBilirubinResult() const
{
  return m_Bilirubin == nullptr ? false : m_Bilirubin->IsValid();
}
SEScalar& SEUrinalysis::GetBilirubinResult()
{
  if (m_Bilirubin == nullptr)
    m_Bilirubin = new SEScalar();
  return *m_Bilirubin;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasSpecificGravityResult() const
{
  return m_SpecificGravity == nullptr ? false : m_SpecificGravity->IsValid();
}
SEScalar& SEUrinalysis::GetSpecificGravityResult()
{
  if (m_SpecificGravity == nullptr)
    m_SpecificGravity = new SEScalar();
  return *m_SpecificGravity;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasBloodResult() const
{
  return m_Blood != SEPresenceIndicator::Invalid;
}
SEPresenceIndicator SEUrinalysis::GetBloodResult() const
{
  return m_Blood;
}
void SEUrinalysis::SetBloodResult(SEPresenceIndicator p)
{
  m_Blood = p;
}
void SEUrinalysis::InvalidateBloodResult()
{
  m_Blood = SEPresenceIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasPHResult() const
{
  return m_pH == nullptr ? false : m_pH->IsValid();
}
SEScalar& SEUrinalysis::GetPHResult()
{
  if (m_pH == nullptr)
    m_pH = new SEScalar();
  return *m_pH;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasProteinResult() const
{
  return m_Protein != SEPresenceIndicator::Invalid;
}
SEPresenceIndicator SEUrinalysis::GetProteinResult() const
{
  return m_Protein;
}
void SEUrinalysis::SetProteinResult(SEPresenceIndicator p)
{
  m_Protein = p;
}
void SEUrinalysis::InvalidateProteinResult()
{
  m_Protein = SEPresenceIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasUrobilinogenResult() const
{
  return m_Urobilinogen == nullptr ? false : m_Urobilinogen->IsValid();
}
SEScalarMassPerVolume& SEUrinalysis::GetUrobilinogenResult()
{
  if (m_Urobilinogen == nullptr)
    m_Urobilinogen = new SEScalarMassPerVolume();
  return *m_Urobilinogen;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasNitriteResult() const
{
  return m_Nitrite != SEPresenceIndicator::Invalid;
}
SEPresenceIndicator SEUrinalysis::GetNitriteResult() const
{
  return m_Nitrite;
}
void SEUrinalysis::SetNitriteResult(SEPresenceIndicator p)
{
  m_Nitrite = p;
}
void SEUrinalysis::InvalidateNitriteResult()
{
  m_Nitrite = SEPresenceIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasLeukocyteEsteraseResult() const
{
  return m_LeukocyteEsterase != SEPresenceIndicator::Invalid;
}
SEPresenceIndicator SEUrinalysis::GetLeukocyteEsteraseResult() const
{
  return m_LeukocyteEsterase;
}
void SEUrinalysis::SetLeukocyteEsteraseResult(SEPresenceIndicator p)
{
  m_LeukocyteEsterase = p;
}
void SEUrinalysis::InvalidateLeukocyteEsteraseResult()
{
  m_LeukocyteEsterase = SEPresenceIndicator::Invalid;
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::HasMicroscopicResult() const
{
  return m_Microscopic != nullptr;
}
SEUrinalysisMicroscopic& SEUrinalysis::GetMicroscopicResult()
{
  if (m_Microscopic == nullptr)
    m_Microscopic = new SEUrinalysisMicroscopic();
  return *m_Microscopic;
}
void SEUrinalysis::RemoveMicroscopicResult()
{
  SAFE_DELETE(m_Microscopic);
}
//-------------------------------------------------------------------------------
bool SEUrinalysis::operator==(SEUrinalysis const& rhs) const
{
  if (this == &rhs)
    return true;

  return ((m_Bilirubin && rhs.m_Bilirubin) ? m_Bilirubin->operator==(*rhs.m_Bilirubin) : m_Bilirubin == rhs.m_Bilirubin)
    && ((m_SpecificGravity && rhs.m_SpecificGravity) ? m_SpecificGravity->operator==(*rhs.m_SpecificGravity) : m_SpecificGravity == rhs.m_SpecificGravity)
    && ((m_pH && rhs.m_pH) ? m_pH->operator==(*rhs.m_pH) : m_pH == rhs.m_pH)
    && ((m_Urobilinogen && rhs.m_Urobilinogen) ? m_Urobilinogen->operator==(*rhs.m_Urobilinogen) : m_Urobilinogen == rhs.m_Urobilinogen)
    && m_Color == rhs.m_Color
    && m_Appearance == rhs.m_Appearance
    && m_Glucose == rhs.m_Glucose
    && m_Ketone == rhs.m_Ketone
    && m_Blood == rhs.m_Blood
    && m_Protein == rhs.m_Protein
    && m_Nitrite == rhs.m_Nitrite
    && m_LeukocyteEsterase == rhs.m_LeukocyteEsterase;
}
bool SEUrinalysis::operator!=(SEUrinalysis const& rhs) const
{
  return !(*this == rhs);
}
//-------------------------------------------------------------------------------
}